Implement security rules for plans (#437)

mrsheepuk opened the pull request:
This PR introduces security rules to the Kore back end, and implements scanning of plans on create/update against those rules. This implements #437. For now, the only rule implemented is a dummy / example rule which will check for whether a plan specifies a wider-than-/16 in the auth proxy allowed IP ranges and raise a warning if not.

UI to view these will be implemented in a separate PR. To view the security scan results, connect to the database container and view the tables security_scan_results and security_rule_results (one scan result contains one-or-more rule results).

bandesz commented on pkg/apis/security/v1/security_scan_types.go:
```diff
@@ -0,0 +1,101 @@
+	// ID is the ID of this scan result in the data store
+	ID uint64 `json:"id,omitempty"`
+	// ResourceAPIVersion is the group and version of the resource scanned by this scan
+	ResourceAPIVersion string `json:"resourceApiVersion,omitempty"`
```
we have an `core/v1/Ownership` type, maybe we could rename that and use that here instead of the 4 Resource* fields?

mrsheepuk replied:
That thought crossed my mind too - I wasn't sure whether there was some overarching problem with re-using it. Essentially both allocations and security scans have a concept of a "reference to another resource of a known type" so it makes sense to generalise. I'll take a look, thanks :)

bandesz commented on pkg/apis/security/v1/security_scan_types.go:
```diff
@@ -0,0 +1,101 @@
+	// ResourceName is the name of the resource scanned by this scan
+	ResourceName string `json:"resourceName,omitempty"`
+	// OwningTeam is the name of the Kore team that owns this resource, will be empty if it is a non-team resource.
+	OwningTeam string `json:"owningTeam,omitempty"`
```
This will be the same as the ResourceNamespace

bandesz replied:
I realised this is the namespace in a team's cluster, so my mistake :)

mrsheepuk replied:
It does depend on the context - at the moment with what resource types we're planning to scan (Cluster and Plan), team will indeed always be the same as namespace, but that might at some point not be true if we start scanning things inside a cluster or similar.

bandesz commented on pkg/persistence/model/migration.go:
```diff
@@ -47,5 +47,10 @@ func Migrations(db *gorm.DB) error {
 		AddForeignKey("team_id", "teams(id)", "CASCADE", "RESTRICT")
 
+	db.AutoMigrate(SecurityScanResult{}).
+		AddIndex("idx_scan_identity", "resource_api_version", "resource_kind", "resource_namespace", "resource_name")
```
shouldn't the resource_kind be the first field in the index?

bandesz replied:
(I know it doesn't matter much, just logically that would be first IMO)

bandesz replied:
A left a comment about this on the PR, so I will remove this.

bandesz commented on pkg/security/authproxyips.go:
```diff
@@ -0,0 +1,115 @@
+
+	var config map[string]interface{}
+	if err := json.Unmarshal(target.Spec.Configuration.Raw, &config); err != nil {
+		result.Message = "Could not check plan as plan JSON invalid"
```
I feel there should be an error state, if there was an internal error with the check (which should be reported back to us)

bandesz replied:
Maybe even better if we extend the rule interface, so it can return an error and we handle that on the scan level.

mrsheepuk replied:
I mused about that - now I'm moving the scanning into a controller, it makes sense for the scanning to report errors normally - when it was happening in-line in the API, I wanted it to be essentially silent if there were problems. Will revise that with the updates I'm making to add the controller.

mrsheepuk commented:
> We should probably use an object exactly like `meta/v1/Ownership` so we don't pass always the same 4-5 values around.

Will do.

> * the index might be better as the following: team, namespace, resource_kind, resource_api_version, resource_name - this way the same index helps looking up all resources for team, a team's namespace or a specific resource

That would prevent the index being used for non-team queries (SQL indexes apply left-to-right so if the first column isn't specified in the query, it wouldn't be able to use the index at all). Adding a second index by team could be useful, though, so I'll do that.

> * Maybe archived_at should be added as the last field of the index, so we can quickly query the latest result

Good idea, will do.

mrsheepuk commented:
@gambol99 @bandesz I've added in the controller implementation and implemented the other suggestions Andras made, and rebased onto master - can you give it a final once over?
